File: problem103.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def zigzagLevelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        answer = []
        turn = True # if true mean left, and false mean right 
        # same idea here -> bfs
        # base case
        if not root:
            return []
        else:
            # add values from left to right
            q = deque() # queue and stack usage for bfs
            q.append(root)
            while q:
                values = []
                totalnodes = len(q) # total nodes in that level
                for _ in range(totalnodes):
                    node = q.popleft()
                    values.append(node.val)
                    if node.left:
                        q.append(node.left)
                    if node.right: 
                        q.append(node.right)
                if not turn:
                    values.reverse()
                answer.append(values)
                turn = not turn # reverse for zigzag input
        return answer

File: problem103.cpp
#include <vector>
#include <queue>
#include <algorithm>
#include <iostream>
using namespace std;
struct TreeNode {
      int val;
      TreeNode *left;
      TreeNode *right;
      TreeNode() : val(0), left(nullptr), right(nullptr) {}
      TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
      TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    vector<vector<int>> zigzagLevelOrder(TreeNode* root) {
        vector<vector<int>> answer;
        bool left_right = true; // check if we reverse it ot not
        // check if tree is not empty or no
        if(!root){
            return answer;
        }
        else{
            queue<TreeNode*> q;
            q.push(root);
            while(!q.empty()){
                int level_size = q.size();
                vector<int> values;
                for(int i = 0; i < level_size; i++){
                    TreeNode* node = q.front(); // pop it from the top
                    q.pop();
                    values.push_back( node -> val);

                    if(node -> left){
                        q.push(node->left);
                    }
                    if(node -> right){
                        q.push(node -> right);
                    }
                }

                if(!left_right){
                    reverse(values.begin(), values.end());
                }
                answer.push_back(values);
                left_right = !left_right;
            }
        }
        return answer;
    }
};